README.md
# Training Processor

A back-end service for picking up a training request, and performing the training.

Overview
--------

The processor monitors a training request queue, once a request is picked up the service requests training and test data from the sensory service.  It fits the model and stores the weights off for later consumption.


High Level Components
---------------------

![Training Processor Diagram](https://github.com/shapeandshare/dicebox.trainingprocessor/raw/master/assets/Training%20Processor%20Diagram.png)


Production Deployment
---------------------

**Docker Container**

The recommended way to run the service is by using the official provided docker container.  The container is published to Docker Hub [here](https://hub.docker.com/r/shapeandshare/dicebox.trainingprocessor/).

The container should be deployed to a Docker Swarm as a service, or as a stand-alone container within a Docker Engine.



**Example**

```
docker service create \
--detach=false \
--replicas 1 \
--log-driver json-file \
--mount type=volume,volume-driver=cloudstor:aws,source=dicebox,destination=/dicebox \
--env SENSORY_SERVER='sensory.yourdomain.com' \
--env TRAINING_PROCESSOR_SERVICE_RABBITMQ_SERVER='rabbitmq.yourdomain.com' \
--env TRAINING_PROCESSOR_SERVICE_RABBITMQ_USERNAME='USERNAME' \
--env TRAINING_PROCESSOR_SERVICE_RABBITMQ_PASSWORD='PASSWORD' \
--env API_ACCESS_KEY='YOUR_API_ACCESS_KEY' \
--env LOAD_BEST_WEIGHTS_ON_START=True \
--env MODEL_WEIGHTS_FILENAME='PREVIOUS_TRAINING_GUID.hdf5' \
--name trainingprocessor shapeandshare/dicebox.trainingprocessor
```


How to apply rolling updates of the service within the swarm:

```
docker service update --image shapeandshare/dicebox.trainingprocessor:latest trainingprocessor
```

In the examples above the Docker Swarm was deployed to AWS and had the Cloudstor:aws plugin enabled and active.
The training service containers will store and read weights from the shared storage.

**Global shared Cloudstor volumes mounted by all tasks in a swarm service.**

The below command is an example of how to create the shared volume within the docker swarm:
```
docker volume create -d "cloudstor:aws" --opt backing=shared dicebox
```

Contributing
------------
1. Fork the [repository](https://github.com/shapeandshare/dicebox.trainingprocessor) on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
MIT License

Copyright (c) 2017-2019 Joshua C. Burt

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

CHANGELOG.md
Change Log
==========

03.09.2019
----------
* Added Change Log!
* Update dicebox.core libraries to the latest.
* Updated requirements.txt managed dependencies to the latest.
* Updated copyright dates.
* Added more exclusions to gitignore.
* Code cleanup and comments added.

03.10.2019
----------
* Updated dicebox.core libraries to address keras weights save/load unicode issue.

03.21.2019
----------
* Don't create a Network File System Connector, we don't have a file system, and not need one, or the categories to be stored to disk.

03.22.2019
----------
* Updated Core Libraries
* Moved category map save location from tmp to weights. (if enabled)

03.23.2019
----------
* Updated Core Libraries to address file descriptor leak when polling for queue consumption
* additional gitignore entries added for dev.

03.30.2019
----------
* Updated core libraries and refactored to support the changes

03.31.2019
----------
* Disabled code which cached data to disk for the training processor until the code to pull from cache is also completed.

04.06.2019
----------
* Refactored to support changes in the core libraries namespace.


app/training_processor.py
#!flask/bin/python
from lib import dicebox_config as config
from lib import sensory_interface
from flask import Flask, jsonify, request, make_response, abort
from flask_cors import CORS, cross_origin
import base64
import logging
import json
from datetime import datetime
import os
import errno
import uuid
import numpy
import pika
#import logging
#import lib.dicebox_config as config
from lib.network import Network
#from datetime import datetime
import json


# Setup logging.
logging.basicConfig(
    format='%(asctime)s - %(levelname)s - %(message)s',
    datefmt='%m/%d/%Y %I:%M:%S %p',
    level=logging.DEBUG,
    filemode='w',
    filename="%s/training_processor.log" % config.LOGS_DIR
)


url = config.TRAINING_PROCESSOR_SERVICE_RABBITMQ_URL
parameters = pika.URLParameters(url)
parameters.heartbeat = 0 # turn this off for now, the timeout otherwise
connection = pika.BlockingConnection(parameters=parameters)
channel = connection.channel()

channel.queue_declare(queue=config.TRAINING_PROCESSOR_SERVICE_RABBITMQ_TRAIN_REQUEST_TASK_QUEUE, durable=True)


# https://stackoverflow.com/questions/273192/how-can-i-create-a-directory-if-it-does-not-exist
def make_sure_path_exists(path):
    try:
        os.makedirs(path)
    except OSError as exception:
        if exception.errno != errno.EEXIST:
            raise


def train_request():
    training_request_id = uuid.uuid4()

    try:
        ## Submit our message
        url = config.TRAINING_SERVICE_RABBITMQ_URL
        logging.debug(url)
        parameters = pika.URLParameters(url)
        connection = pika.BlockingConnection(parameters=parameters)

        channel = connection.channel()

        channel.queue_declare(queue=config.TRAINING_SERVICE_RABBITMQ_TRAIN_REQUEST_TASK_QUEUE, durable=True)

        training_request = {}
        training_request['training_request_id'] = str(training_request_id)
        channel.basic_publish(exchange=config.TRAINING_SERVICE_RABBITMQ_EXCHANGE,
                              routing_key=config.TRAINING_SERVICE_RABBITMQ_TRAINING_REQUEST_ROUTING_KEY,
                              body=json.dumps(training_request),
                              properties=pika.BasicProperties(
                                  delivery_mode=2,  # make message persistent
                              ))
        logging.debug(" [x] Sent %r" % json.dumps(training_request))
        connection.close()
    except:
        # something went wrong..
        logging.error('we had a failure sending the request to the message system')
        return None

    return training_request_id


def train_call(training_request_id):
    logging.debug("processing training request id: (%s)" % training_request_id)
    network = Network(config.NN_PARAM_CHOICES)
    if config.LOAD_BEST_WEIGHTS_ON_START is True:
        logging.info('attempting to restart training from previous session..')
        network.create_lonestar(create_model=True,
                                weights_filename="%s/%s" % (config.WEIGHTS_DIR, config.MODEL_WEIGHTS_FILENAME))
    else:
        network.create_lonestar(create_model=True)

    with open('./category_map.json', 'w') as category_mapping_file:
        category_mapping_file.write(json.dumps(network.fsc.CATEGORY_MAP))

    i = 1
    while i <= config.EPOCHS:
        logging.info("epoch (%i of %i)" % (i, config.EPOCHS))
        network.train_and_save(config.DATASET)

        # save the model after every epoch, regardless of accuracy.
        #path = "%s/%s/" % (config.WEIGHTS_DIR, training_request_id)
        #make_sure_path_exists(path)
        #filename = "weights.epoch_%i.final.%s.hdf5" % (i, datetime.now().strftime('%Y-%m-%d_%H_%M_%S_%f'))

        #full_path = "%s%s" % (path, filename)
        #logging.info("saving model weights after epoch %i to file %s" % (i, full_path))
        #network.save_model(filename)

        make_sure_path_exists(config.WEIGHTS_DIR)
        full_path = "%s/%s.hdf5" % (config.WEIGHTS_DIR, training_request_id)
        logging.info("saving model weights after epoch %i to file %s" % (i, full_path))
        network.save_model(full_path)

        # the next epoch..
        i += 1

    logging.info("network accuracy: %.2f%%" % (network.accuracy * 100))
    logging.info('-'*80)
    network.print_network()

    return None





print(' [*] Waiting for messages. To exit press CTRL+C')

def callback(ch, method, properties, body):
    logging.info(" [x] Received %r" % body)
    ch.basic_ack(delivery_tag=method.delivery_tag)
    train_call(json.loads(body)['training_request_id'])
    logging.info(" [x] Done")

channel.basic_qos(prefetch_count=1)
channel.basic_consume(callback,
                      queue=config.TRAINING_PROCESSOR_SERVICE_RABBITMQ_TRAIN_REQUEST_TASK_QUEUE)

channel.start_consuming()

app/trainingprocessor.py
#!flask/bin/python
###############################################################################
# Training Processor
#
# Copyright (c) 2017-2019 Joshua Burt
###############################################################################


###############################################################################
# Dependencies
###############################################################################
import logging
import os
import errno
import pika
import json
from dicebox.config.dicebox_config import DiceboxConfig
from dicebox.dicebox_network import DiceboxNetwork

# Config
config_file = './dicebox.config'
CONFIG = DiceboxConfig(config_file)


###############################################################################
# Allows for easy directory structure creation
# https://stackoverflow.com/questions/273192/how-can-i-create-a-directory-if-it-does-not-exist
###############################################################################
def make_sure_path_exists(path):
    try:
        if os.path.exists(path) is False:
            os.makedirs(path)
    except OSError as exception:
        if exception.errno != errno.EEXIST:
            raise


###############################################################################
# Setup logging.
###############################################################################
make_sure_path_exists(CONFIG.LOGS_DIR)
logging.basicConfig(
    format='%(asctime)s - %(levelname)s - %(message)s',
    datefmt='%m/%d/%Y %I:%M:%S %p',
    level=logging.DEBUG,
    filemode='w',
    filename="%s/trainingprocessor.%s.log" % (CONFIG.LOGS_DIR, os.uname()[1])
)

# https://github.com/pika/pika/issues/692
# reduce log level of pika
logging.getLogger("pika").setLevel(logging.WARNING)

###############################################################################
# Message System Configuration
###############################################################################
url = CONFIG.TRAINING_PROCESSOR_SERVICE_RABBITMQ_URL
logging.debug('-'*80)
logging.debug("rabbitmq url: (%s)", url)
logging.debug('-'*80)

parameters = pika.URLParameters(url)
parameters.heartbeat = 0  # turn this off for now, the timeout otherwise
connection = pika.BlockingConnection(parameters=parameters)
channel = connection.channel()

channel.queue_declare(queue=CONFIG.TRAINING_PROCESSOR_SERVICE_RABBITMQ_TRAIN_REQUEST_TASK_QUEUE, durable=True)
channel.basic_qos(prefetch_count=0)


###############################################################################
# Training Logic
###############################################################################
def train_call(training_request_id):
    logging.debug('-' * 80)
    logging.debug("processing training request id: (%s)" % training_request_id)
    logging.debug('-' * 80)
    network = DiceboxNetwork(CONFIG.NN_PARAM_CHOICES, True)
    if CONFIG.LOAD_BEST_WEIGHTS_ON_START is True:
        logging.debug('-' * 80)
        logging.debug('attempting to restart training from previous session..')
        logging.debug('-' * 80)
        network.create_lonestar(create_model=True,
                                weights_filename="%s/%s" % (CONFIG.WEIGHTS_DIR, CONFIG.MODEL_WEIGHTS_FILENAME))
        logging.debug('-' * 80)
        logging.debug('Done')
        logging.debug('-' * 80)
    else:
        logging.debug('-' * 80)
        logging.debug('creating model, but NOT loading previous weights.')
        logging.debug('-' * 80)
        network.create_lonestar(create_model=True)
        logging.debug('-' * 80)
        logging.debug('Done')
        logging.debug('-' * 80)

    if network.fsc is not None:
        logging.debug('-' * 80)
        logging.debug('writing category map to %s for later use with the weights.', CONFIG.TMP_DIR)
        logging.debug('-' * 80)
        with open('%s/category_map.json' % CONFIG.WEIGHTS_DIR , 'w') as category_mapping_file:
            category_mapping_file.write(json.dumps(network.fsc.category_map))

    i = 1
    while i <= CONFIG.EPOCHS:
        logging.debug('-' * 80)
        logging.debug("epoch (%i of %i)" % (i, CONFIG.EPOCHS))
        logging.debug('-' * 80)
        network.train_and_save(CONFIG.DATASET)

        make_sure_path_exists(CONFIG.WEIGHTS_DIR)
        logging.debug('-' * 80)
        full_path = "%s/%s.%.2f.hdf5" % (CONFIG.WEIGHTS_DIR, training_request_id, (network.accuracy * 100))
        logging.debug("saving model weights after epoch %i to file %s" % (i, full_path))
        logging.debug('-' * 80)
        network.save_model(full_path)

        # the next epoch..
        i += 1

    logging.debug('-' * 80)
    logging.debug("network accuracy: %.2f%%" % (network.accuracy * 100))
    logging.debug('-'*80)

    network.print_network()
    logging.debug(network.print_network())
    logging.debug('-' * 80)
    return None


###############################################################################
# Our callback when message consumption is ready to occur
###############################################################################
def callback(ch, method, properties, body):
    logging.debug('-' * 80)
    logging.debug(" [x] Received %r" % body)
    logging.debug('-' * 80)
    ch.basic_ack(delivery_tag=method.delivery_tag)
    train_call(json.loads(body)['training_request_id'])
    logging.debug('-' * 80)
    logging.debug(" [x] Done")
    logging.debug('-'*80)


###############################################################################
# Main wait loop begins now ..
###############################################################################
logging.debug('-'*80)
print(' [*] Waiting for messages. To exit press CTRL+C')
logging.debug('-'*80)
channel.basic_consume(callback,
                      queue=CONFIG.TRAINING_PROCESSOR_SERVICE_RABBITMQ_TRAIN_REQUEST_TASK_QUEUE)
channel.start_consuming()
